expanded weights: instance norm faster rules

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70141

Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

Commit:
expanded weights faster rules

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"




[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"




[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, LayerNorm, and Linear.

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, LayerNorm, and Linear.

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, LayerNorm, and Linear.

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2289,6 +2289,8 @@
 
 - func: group_norm(Tensor input, int num_groups, Tensor? weight=None, Tensor? bias=None, float eps=1e-05, bool cudnn_enabled=True) -> Tensor
 
+- func: _group_norm_all_outputs(Tensor input, int num_groups, Tensor? weight=None, Tensor? bias=None, float eps=1e-05) -> (Tensor, Tensor, Tensor)
+
```
Why do we need a new operator? Is it to save values so that we can compute the gradients of the input Tensor?

zou3519 replied:
Also, what is the difference between _group_norm_all_outputs and native_group_norm?

samdow replied:
Yes exactly to save the mean and variance so we can compute the gradients of the input tensor.

Difference between `_group_norm_all_outputs` and `native_group_norm` is that `_group_norm_all_outputs` will do [some validation](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/group_norm.cpp#L131-L152) (originally done in `group_norm`, now `group_norm` just calls `_group_norm_all_outputs`) before calling `native_group_norm`

samdow replied:
cc @albanD How do you feel about the functions added to native_funcitons? Any alternative ideas?

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

[ghstack-poisoned]

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34350951).

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34350951).

zou3519 reviewed (commented):
This looks good to me, I didn't read through the math but I assume it's all pulled from Opacus.

I want to get at least one more person's take on if it's OK to add these additional operators to native_functions.yaml. The alternative seems likes it would be to just rewrite them in Python, but that doesn't seem very good because it involves more code duplication :/

samdow commented:
Sounds good--I think I figured out what the failures are from (most of the non-libtorch look real) so I'll work on fixing those in the meantime

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: layer norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: layer norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: layer norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: layer norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: layer norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34350951).

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

samdow commented:
@samdow has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34350951).

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

samdow commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -606,8 +606,16 @@ Tensor instance_norm(
   Tensor running_var_ = repeat_if_defined(running_var, b);
 
   auto input_reshaped = input.contiguous().view(shape);
-  auto out = at::batch_norm(input_reshaped, weight_, bias_, running_mean_, running_var_,
-                            use_input_stats, momentum, eps, cudnn_enabled);
```
revert

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

samdow commented:
This version uses `native_batch_norm_backward` always because it works for all inputs and the metadata about which (naive, miopen, or cuda) function was used in the forward is no longer saved. This means that we don't add a new native_function entry though

Doing some quick perf testing, 1000 layers of instance norm, ~1000 iterations of batch size of 64 run in ~304s using expanded weights and ~341s using Opacus on an A100 so it looks like any performance drop from using `native_batch_norm_backward` isn't very significant (or may be offset by the win in not running the entire backwards pass)

samdow commented:
@pytorchbot merge this please

Commit:
Update on "expanded weights: instance norm faster rules"


Introduces faster per sample gradient rules for the functional versions of Conv2d, Embedding, GroupNorm, InstanceNorm, and LayerNorm

Description of overall design and how we register per sample gradient rules [here](https://gist.github.com/samdow/fa0a164fec7963f93ff45284989cfc55)

Differential Revision: [D34350951](https://our.internmc.facebook.com/intern/diff/D34350951)

[ghstack-poisoned]

samdow commented:
@pytorchbot merge this please
